tensorflow_io/parquet/kernels/parquet_dataset_ops.cc

```diff
@@ -0,0 +1,331 @@
+        ctx, filenames_tensor->dims() <= 1,
+        errors::InvalidArgument("`filenames` must be a scalar or a vector."));
+
+    const Tensor* columns_tensor;
+    OP_REQUIRES_OK(ctx, ctx->input("columns", &columns_tensor));
+    OP_REQUIRES(
+        ctx, columns_tensor->dims() <= 1,
+        errors::InvalidArgument("`columns` must be a scalar or a vector."));
+
+    std::vector<string> filenames;
+    filenames.reserve(filenames_tensor->NumElements());
+    for (int i = 0; i < filenames_tensor->NumElements(); ++i) {
+      filenames.push_back(filenames_tensor->flat<string>()(i));
+    }
+
+    std::vector<int64> columns;
+    columns.reserve(columns_tensor->NumElements());
+    for (int i = 0; i < columns_tensor->NumElements(); ++i) {
+      columns.push_back(columns_tensor->flat<int32>()(i));
+    }
+
+    *output =
+        new Dataset(ctx, filenames, columns, output_types_, output_shapes_);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    Dataset(OpKernelContext* ctx, const std::vector<string>& filenames,
+            const std::vector<int64>& columns,
+            const DataTypeVector& output_types,
+            const std::vector<PartialTensorShape>& output_shapes)
+        : DatasetBase(DatasetContext(ctx)),
+          filenames_(filenames),
+          columns_(columns),
+          output_types_(output_types),
+          output_shapes_(output_shapes) {}
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::Parquet")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return output_types_;
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return output_shapes_;
+    }
+
+    string DebugString() const override { return "ParquetDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      Node* filenames = nullptr;
+      TF_RETURN_IF_ERROR(b->AddVector(filenames_, &filenames));
+      Node* columns = nullptr;
+      TF_RETURN_IF_ERROR(b->AddVector(columns_, &columns));
+      TF_RETURN_IF_ERROR(b->AddDataset(this, {filenames, columns}, output));
+      return Status::OK();
+    }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        mutex_lock l(mu_);
+
+        // Loop until we find a row to read or there are no more files left to
+        // read
+        while (true) {
+          if (parquet_reader_) {
```

Comment:
minor: consider inverting the if condition (`if (!parquet_reader_)`) to reduce indentation?

Reply:
Also to further reduce indentation, should we refactor some code into a helper function, say a helper to read all row groups from the current file (the body of the while loop below)?